### PR TITLE
Feature/pull to refresh

### PR DESCRIPTION
**Pull request by djkwagala**
This is the initial pull request

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,19 @@
 # pull-to-refresh
+Swipe or pull down to reload a page
```
Comment (Andries-Smit) [inline README.md]: no swipe

File: LICENSE [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "{}"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright {yyyy} {name of copyright owner}
```
Comment (Andries-Smit) [inline LICENSE]: please fill in

File: README.md [hunk 3]
```diff
@@ -1,2 +1,19 @@
...
+
+## Features
+* Pull down to refresh a page
+
```
Comment (Andries-Smit) [inline README.md]: needs more info.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "pulltorefresh",
+  "version": "0.1.0",
+  "description": "Change a number value using a slider",
```
Comment (Andries-Smit) [inline package.json]: ?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  "scripts": {
```
Comment (Andries-Smit) [inline package.json]: If there are no tests. remove all test files and reference like karma enzyme etc.

File: package.json [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    "source-map-loader": "^0.1.5",
+    "style-loader": "^0.13.1",
+    "ts-loader": "^1.3.3",
+    "tslint": "^4.3.1",
+    "tslint-eslint-rules": "^3.2.3",
+    "typescript": "^2.1.5",
+    "webpack": "^1.13.2",
+    "webpack-dev-server": "^1.16.2"
+  },
+  "dependencies": {
```
Comment (Andries-Smit) [inline package.json]: no longer....

File: src/com/mendix/widget/PullToRefresh/PullToRefresh.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        this.pullToRefresh = new PullToRefresh({
+            mainElement: document.body,
+            pullToRefreshElement: this.pullToRefreshElement
+        });
+
+        this.pullToRefresh.setupEvents();
+    }
+
+    uninitialize() {
+        this.pullToRefresh.removeEvents();
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/PullToRefresh.ts]: Why not removing the dom too?

File: src/com/mendix/widget/PullToRefresh/PullToRefresh.xml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<widget id="com.mendix.widget.PullToRefresh.PullToRefresh" needsEntityContext="false" xmlns="http://www.mendix.com/widget/1.0/">
+    <name>PullToRefresh</name>
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/PullToRefresh.xml]: Add spaces `Pull to refresh`

File: src/com/mendix/widget/PullToRefresh/PullToRefresh.xml [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    <description>Swipe or pull down to reload a page</description>
+    <icon/>
+    <properties>
+        <property key="key1" type="boolean" defaultValue="false">
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/PullToRefresh.xml]: what does this do?

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+        this.onReset = this.onReset.bind(this);
+
+        this.settings = {
+            classPrefix: "pull-to-refresh-",
+            cssProp: "min-height",
+            iconArrow: "&#8675;",
+            iconRefreshing: "&hellip;",
+            mainElement: params.mainElement,
+            maximumDistance: 150,
+            onRefresh: () => { mx.ui.reload(); },
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: move to constructor param.
The whole file is a generic pull to refresh class. The `reload` is mx specific

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        };
+        this.state = "pending";
+        this.pullStart = { screenX: 0, screenY: 0 };
+    }
+
+    setupEvents() {
+        window.addEventListener("touchstart", this.onTouchStart);
+        window.addEventListener("touchend", this.onTouchEnd);
+        (window.addEventListener as WhatWGAddEventListener)("touchmove", this.onTouchMove, { passive: false });
+        this.setup = true;
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: private attribute is never used, if not used, please remove

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+            pullToRefreshElement,
+            iconArrow,
+            iconRefreshing,
+            refreshText,
+            pullToRefreshText,
+            releaseToRefreshText
+        } = this.settings;
+
+        const iconElement = pullToRefreshElement.querySelector(`.${classPrefix}icon`);
+        const textElement = pullToRefreshElement.querySelector(`.${classPrefix}text`);
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Query selectors should are always returning the same node. so they should not be done during update, move them to the constructing/setup phase

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+
+    removeEvents() {
+        window.removeEventListener("touchstart", this.onTouchStart);
+        window.removeEventListener("touchend", this.onTouchEnd);
+        (window.removeEventListener as WhatWGAddEventListener)("touchmove", this.onTouchMove, { passive: false });
+        this.setup = false;
+    }
+
+    private update() {
+        const {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: if cleaned up, no need for desctructoring

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+            this.update();
+        }
+        if (this.pullStart.screenY && this.pullMoveY) {
+            this.distance = this.pullMoveY - this.pullStart.screenY;
+        }
+        if (this.distance > 0 && Math.abs(this.lastDistance - this.distance) > 1) {
+            this.lastDistance = this.distance;
+            event.preventDefault();
+            domStyle.set(pullToRefreshElement, cssProp, `${this.distanceResisted}px`);
+            this.distanceResisted = this.settings.resistanceFunction(this.distance / thresholdDistance)
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Calculation is done after the value is set. Why not use `const`

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+            this.state = "pulling";
+            this.update();
+        }
+        if (this.pullStart.screenY && this.pullMoveY) {
+            this.distance = this.pullMoveY - this.pullStart.screenY;
+        }
+        if (this.distance > 0 && Math.abs(this.lastDistance - this.distance) > 1) {
+            this.lastDistance = this.distance;
+            event.preventDefault();
+            domStyle.set(pullToRefreshElement, cssProp, `${this.distanceResisted}px`);
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Dont set style, if style has the same value als last time

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+            this.distanceResisted = this.settings.resistanceFunction(this.distance / thresholdDistance)
+                * Math.min(maximumDistance, this.distance);
+
+            if (this.state === "pulling" && this.distanceResisted > thresholdDistance) {
+                domClass.add(pullToRefreshElement.id, `${classPrefix}release`);
+                this.state = "releasing";
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Name if state "releasing" does not cover the real state. `will refresh on release` ......

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+                this.update();
+            }
+
+            if (this.state === "releasing" && this.distanceResisted < thresholdDistance) {
+                domClass.remove(pullToRefreshElement.id, `${classPrefix}release`);
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Move this dom manipulation to update

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+        }
+        if (this.distance > 0 && Math.abs(this.lastDistance - this.distance) > 1) {
+            this.lastDistance = this.distance;
+            event.preventDefault();
+            domStyle.set(pullToRefreshElement, cssProp, `${this.distanceResisted}px`);
+            this.distanceResisted = this.settings.resistanceFunction(this.distance / thresholdDistance)
+                * Math.min(maximumDistance, this.distance);
+
+            if (this.state === "pulling" && this.distanceResisted > thresholdDistance) {
+                domClass.add(pullToRefreshElement.id, `${classPrefix}release`);
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Move this dom manipulation to update

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+                this.state = "pulling";
+                this.update();
+            }
+        }
+    }
+
+    private onTouchEnd() {
+        const { pullToRefreshElement, onRefresh, cssProp, classPrefix } = this.settings;
+
+        if (this.state === "releasing" && this.distanceResisted > this.settings.thresholdDistance) {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Check for `this.distanceResisted > this.settings.thresholdDistance` not needed, `this.state` can only be ``releasing` when that condition is true, right?

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+            }, this.settings.refreshTimeout);
+        } else {
+            if (this.state === "refreshing") return;
+
+            domStyle.set(pullToRefreshElement, cssProp, "0px");
+            this.state = "pending";
+        }
+
+        this.update();
+        domClass.remove(pullToRefreshElement.id, `${classPrefix}release ${classPrefix}pull`);
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Dom manupulation should be moved to update

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+            return;
+        }
+
+        clearTimeout(this.timeout);
+
+        this.pullStart.screenX = event.touches[0].screenX;
+        this.pullStart.screenY = event.touches[0].screenY;
+        this.enable = this.settings.triggerElement.contains(event.target as Node);
+        this.state = "pending";
+        this.lastDistance = 0;
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: also set `this.distanceResisted = 0;`

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+            if (this.state === "releasing") {
+                textElement.innerHTML = releaseToRefreshText;
+            }
+            if (this.state === "pulling" || this.state === "pending") {
+                textElement.innerHTML = pullToRefreshText;
+            }
+        }
+    }
+
+    private onReset() {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: name does not cover what the function is doing.  `resetDom` ???

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+                const afterRefresh = onRefresh(this.onReset);
+                if (afterRefresh && typeof afterRefresh.then === "function") {
+                    afterRefresh.then(() => this.onReset());
+                }
+                if (!afterRefresh && !onRefresh.length) {
+                    this.onReset();
+                }
+            }, this.settings.refreshTimeout);
+        } else {
+            if (this.state === "refreshing") return;
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: strange if else construction.
`if ()[} else if (){} else`

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+
+            domStyle.set(pullToRefreshElement, cssProp, "0px");
+            this.state = "pending";
+        }
+
+        this.update();
+        domClass.remove(pullToRefreshElement.id, `${classPrefix}release ${classPrefix}pull`);
+        this.pullStart.screenY = this.pullStart.screenX = this.pullMoveY = 0;
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Why reset, this is already done in the start?
`this.pullStart = { screenX: 0, screenY: 0 };` looks cleaner

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+                    afterRefresh.then(() => this.onReset());
+                }
+                if (!afterRefresh && !onRefresh.length) {
+                    this.onReset();
+                }
+            }, this.settings.refreshTimeout);
+        } else {
+            if (this.state === "refreshing") return;
+
+            domStyle.set(pullToRefreshElement, cssProp, "0px");
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: why not just call the `onReset` function?

File: typings/PullToRefresh.d.ts [hunk 26]
```diff
@@ -0,0 +1,35 @@
+declare module "react-pull-to-refresh" {
```
Comment (Andries-Smit) [inline typings/PullToRefresh.d.ts]: Is this file still needed?

File: typings/react.d.ts [hunk 27]
```diff
@@ -0,0 +1,8 @@
+import { ClassAttributes, DOMAttributes, ReactNode, DOMElement } from "react";
```
Comment (Andries-Smit) [inline typings/react.d.ts]: is this file still needed with latest react typing?

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts [hunk 28]
```diff
@@ -0,0 +1,215 @@
...
+    }
+
+    removeEvents() {
+        window.removeEventListener("touchstart", this.onTouchStart);
+        window.removeEventListener("touchend", this.onTouchEnd);
+        (window.removeEventListener as WhatWGAddEventListener)("touchmove", this.onTouchMove, { passive: false });
+        this.setup = false;
+    }
+
+    private update() {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Maybe this function should have a parameter `nextState` an `position` so it can handel all dom manupulations

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 15 above)
Comment (Andries-Smit) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: transform is probably better performant

**Review by Andries-Smit — COMMENTED** [review]
Code requires some more improvements

File: README.md (hunk 1 above)
Comment (akileng56) [reply README.md]: Removed

File: README.md (hunk 3 above)
Comment (akileng56) [reply README.md]: Did you mean adding more to that feature?
Modified  it to: Pull down page until the text 'release to refresh' appears in order to refresh the page

File: package.json (hunk 4 above)
Comment (akileng56) [reply package.json]: changed the description

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 10 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: moved

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 11 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: removed since its unnecessary

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 28 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Added

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 13 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: improved

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 12 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: added to constructor

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 22 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Done

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 21 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Done

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 18 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: moved

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 17 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: moved

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 19 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: true

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 23 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: strange indeed

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 25 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Done. thanks

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 20 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Improved the function and moved this to resetDom function

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 24 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: added to resetDom function. (sometimes onTouchStart is not executed like on my device) so having it all in reset is better.

File: typings/PullToRefresh.d.ts (hunk 26 above)
Comment (djkwagala) [reply typings/PullToRefresh.d.ts]: Removed

File: src/com/mendix/widget/PullToRefresh/PullToRefresh.xml (hunk 8 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/PullToRefresh.xml]: Done

File: src/com/mendix/widget/PullToRefresh/PullToRefresh.ts (hunk 7 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/PullToRefresh.ts]: Done

File: package.json (hunk 6 above)
Comment (djkwagala) [reply package.json]: Removed

File: LICENSE (hunk 2 above)
Comment (djkwagala) [reply LICENSE]: Done

File: typings/react.d.ts (hunk 27 above)
Comment (djkwagala) [reply typings/react.d.ts]: removed

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 15 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: I've used direct style attribute manipulation. `pullToRefreshElement.style.minHeight = '${this.distanceResisted}px` I think its much cheaper on resources than both.

File: src/com/mendix/widget/PullToRefresh/PullToRefresh.xml (hunk 9 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/PullToRefresh.xml]: Fixed

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 16 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: state `releaseToRefresh`

File: src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts (hunk 14 above)
Comment (djkwagala) [reply src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts]: Done.

**Comment by djkwagala** [comment]
I've tested the offline sync, its working well. This is ready for external review and delivery to mendix
